core/tests/unit/test_client.py

```diff
@@ -221,3 +234,47 @@ def test_ctor_explicit_bytes(self):
     def test_ctor_explicit_unicode(self):
         PROJECT = u'PROJECT'
         self._explicit_ctor_helper(PROJECT)
+
+    def _from_service_account_json_helper(self, project=None):
+        from google.cloud import _helpers
+
+        klass = self._get_target_class()
+
+        info = {'dummy': 'value', 'valid': 'json'}
+        if project is None:
+            expected_project = 'eye-d-of-project'
+        else:
+            expected_project = project
+
+        info['project_id'] = expected_project
+        # Mock both the file opening and the credentials constructor.
+        json_fi = io.StringIO(_helpers._bytes_to_unicode(json.dumps(info)))
+        file_open_patch = mock.patch(
+            'io.open', return_value=json_fi)
+        constructor_patch = mock.patch(
+            'google.oauth2.service_account.Credentials.'
+            'from_service_account_info',
+            return_value=_make_credentials())
+
+        with file_open_patch as file_open:
+            with constructor_patch as constructor:
+                kwargs = {}
+                if project is not None:
+                    kwargs['project'] = project
+                client_obj = klass.from_service_account_json(
+                    mock.sentinel.filename, **kwargs)
+
+        self.assertIs(
+            client_obj._credentials, constructor.return_value)
+        self.assertIsNone(client_obj._http_internal)
+        self.assertEqual(client_obj.project, expected_project)
+        # Check that mocks were called as expected.
+        file_open.assert_called_once_with(
+            mock.sentinel.filename, 'r', encoding='utf-8')
+        constructor.assert_called_once_with(info)
```

Comment:
I can factor out the common parts? The only difference here is the `project` / `project_id`